src/java/main/org/apache/zookeeper/server/quorum/QuorumCnxManager.java

```diff
@@ -880,19 +876,20 @@ public void run() {
          * 
          * @return boolean  Value of variable running
          */
-        synchronized boolean finish() {
-            if(!running){
-                /*
-                 * Avoids running finish() twice. 
-                 */
-                return running;
-            }
-            running = false;            
-
-            this.interrupt();
-            threadCnt.decrementAndGet();
-            return running;
-        }
+		boolean finish() {
```

Comment:
I understand what you are trying to do here and why this is optimized. I'm just concerned that that any performance increase here would not be worth the decrease in readability.